### PR TITLE
BigTreeTech E3 RRF board (STM32F407VGT6)

### PR DESCRIPTION
**Pull request by thinkyhead**
A new board with the same processor as the GTR 1.0 but in a form factor to fit the Ender 3.

Some changes related to SDIO were also made in the BTT-E3-RRF branch, so those may also be added here in a separate commit for evaluation.

### CODE REVIEW

**Comment by thisiskeithb** [comment]
This will resolve https://github.com/MarlinFirmware/Marlin/issues/21580

File: Marlin/src/core/boards.h [hunk 1]
```diff
@@ -366,21 +366,22 @@
...
-#define BOARD_FYSETC_S6_V2_0          4216  // FYSETC S6 v2.0 (STM32F446VET6)
-#define BOARD_FYSETC_SPIDER           4217  // FYSETC Spider (STM32F446VET6)
-#define BOARD_FLYF407ZG               4218  // FLYF407ZG (STM32F407ZG)
-#define BOARD_MKS_ROBIN2              4219  // MKS_ROBIN2 (STM32F407ZE)
-#define BOARD_MKS_ROBIN_PRO_V2        4220  // MKS Robin Pro V2 (STM32F407VE)
-#define BOARD_MKS_ROBIN_NANO_V3       4221  // MKS Robin Nano V3 (STM32F407VG)
-#define BOARD_ANET_ET4                4222  // ANET ET4 V1.x (STM32F407VGT6)
-#define BOARD_ANET_ET4P               4223  // ANET ET4P V1.x (STM32F407VGT6)
-#define BOARD_FYSETC_CHEETAH_V20      4224  // FYSETC Cheetah V2.0
+#define BOARD_BTT_E3_RRF              4210  // BigTreeTech GTR v1.0 (STM32F407IGT)
```
Comment (marcinwisniowski) [inline Marlin/src/core/boards.h]: Comment should be updated to
`#define BOARD_BTT_E3_RRF 4210 // BigTreeTech E3 RRF (STM32F407IGT)`

Comment (thinkyhead) [reply Marlin/src/core/boards.h]: STM32F4075VX

**Comment by thinkyhead** [comment]
Changes in the STM32 HAL for SDIO were made to Marlin on March 16 by @Msq001. These changes conflict (or at least differ in detail) significantly from the changes made on March 9 by @rhapsodyv.

So, I've disabled our current version of the code with some `#if 0` blocks and activated the BTT code here for testing. But ultimately @rhapsodyv and @Msq001 will be able to tell us whether the changes they each made to the code differ in any meaningful ways.

**Comment by thinkyhead** [comment]
It looks like the changes for SDIO are too STM32F4-centric to compile on other boards. Maybe some of those STM32F4-specific changes will be useful, but they'll have to be conditionalized.

**Comment by bigtreetech** [comment]
> Changes in the STM32 HAL for SDIO were made to Marlin on March 16 by @Msq001. These changes conflict (or at least differ in detail) significantly from the changes made on March 9 by @rhapsodyv.
> 
> So, I've disabled our current version of the code with some `#if 0` blocks and activated the BTT code here for testing. But ultimately @rhapsodyv and @Msq001 will be able to tell us whether the changes they each made to the code differ in any meaningful ways.

After https://github.com/MarlinFirmware/Marlin/pull/21476, STM32F4 should also work normally, and it should also work on E3-RRF and SKR-2.
So the code of SDIO need not be changed.

**Comment by rhapsodyv** [comment]
Yes. I took insights from @bigtreetech to fix the SDIO.

My fix was tested and working in a F103, F401 and a F407.

Probably it will work without any change in this new btt board.

**Comment by rhapsodyv** [comment]
> It looks like the changes for SDIO are too STM32F4-centric to compile on other boards. Maybe some of those STM32F4-specific changes will be useful, but they'll have to be conditionalized.

Yes, the changes from this PR are _F4 only_.... My changes are for F1 and F4. You should revert the changes and keep my original code.
If it don't work in this board, we should test and fix what is wrong.

File: Marlin/src/core/boards.h [hunk 2]
```diff
@@ -366,21 +366,22 @@
...
-#define BOARD_FYSETC_S6_V2_0          4216  // FYSETC S6 v2.0 (STM32F446VET6)
-#define BOARD_FYSETC_SPIDER           4217  // FYSETC Spider (STM32F446VET6)
-#define BOARD_FLYF407ZG               4218  // FLYF407ZG (STM32F407ZG)
-#define BOARD_MKS_ROBIN2              4219  // MKS_ROBIN2 (STM32F407ZE)
-#define BOARD_MKS_ROBIN_PRO_V2        4220  // MKS Robin Pro V2 (STM32F407VE)
-#define BOARD_MKS_ROBIN_NANO_V3       4221  // MKS Robin Nano V3 (STM32F407VG)
-#define BOARD_ANET_ET4                4222  // ANET ET4 V1.x (STM32F407VGT6)
-#define BOARD_ANET_ET4P               4223  // ANET ET4P V1.x (STM32F407VGT6)
-#define BOARD_FYSETC_CHEETAH_V20      4224  // FYSETC Cheetah V2.0
+#define BOARD_BTT_E3_RRF              4210  // BigTreeTech E3 RRF (STM32F4075VX)
```
Comment (thisiskeithb) [inline Marlin/src/core/boards.h]: This is an STM32F407VGT6-based board.

Comment (rhapsodyv) [reply Marlin/src/core/boards.h]: I'm almost sure there's no `STM32F4075VX` mcu :-p ... it's 407 or 405... :-p

File: Marlin/src/pins/stm32f4/pins_BTT_E3_RRF.h [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+//
+#define TEMP_BED_PIN                        PA1   // Analog Input "TB"
+#define TEMP_0_PIN                          PA0   // Analog Input "TH0"
+
+//
+// Heaters / Fans
+//
+#define HEATER_BED_PIN                      PB4   // "HB"
+#define HEATER_0_PIN                        PB3   // "HE0"
+#define HEATER_1_PIN                        PE4   // "HE1"
```
Comment (ondrada) [inline Marlin/src/pins/stm32f4/pins_BTT_E3_RRF.h]: Line 170 (Heater 1 Pin) should be moved to somewhere around line 73 (it  is on IDEX module) and should look like 
`#define HEATER_1_PIN                        FPC16_PIN   // "HE1"`

File: Marlin/src/core/boards.h (hunk 2 above)
Comment (thinkyhead) [reply Marlin/src/core/boards.h]: Righto. So the platform's name is basically a mishmash with X as a placeholder, not a specific MCU. Gotcha.

**Comment by rhapsodyv** [comment]
We should try to make this board use the generic variant too... or least one of the current f4 variant!!

We should stop copying and pasting a bunch of 99% equal files for each added board... this is non sense and will lead to state we can call: variant hell... 

There's already about 4 or 5 F407 variants.

I really doubt that no one will work with this board.

**Comment by thinkyhead** [comment]
Let's go step-by-step how to convert the **changes** made by modifying the variant into **flags** that will be picked up by the generic variant:

1. `THE_BOARD.json`
  - Specified using environment `board = THE_BOARD`
  - Contains values otherwise set by environment `board_build.fieldname =`
2. `variants/THE_VARIANT` folder
  - Specified by `board_build.variant = THE_VARIANT`,
  - ...or in `THE_BOARD.json` by `{ "build": { "variant":"THE_VARIANT" } }`.
3. `hal_conf_extra.h` — Flags to enable platform HAL features
4. `ldscript.ld`
5. `PeripheralPins.c`
6. `PinNamesVar.h`
7. `variant.cpp`
8. `variant.h`

**Comment by rhapsodyv** [comment]
### 1. `THE_BOARD.json`

We just need the `THE_BOARD.json` when there's no generic board available for this mcu. Or if the current generic_board.json from PIO have some bug.

For this board, we can just use: `board = genericSTM32F407VGT6` and configure any extra build flag needed in the `build_flag` option.

Everything other board.json that is copied/added to Marlin should have a strong reason. For example, I copied `genericSTM32F407VGT6.json` to marlin because the original has a bug. I sent a PR to PIO and they merged my board fix. But until PIO release a new version, we need use the copy.

Current available boards: `https://github.com/platformio/platform-ststm32/tree/develop/boards`

### 2. `hal_conf_extra.h`

We don't need it either. Marlin have some standard features that are enabled by default in the generic variant. If the new board needs any other module, it can enable just adding `-DHAL_MODULE_FOO` to the `build_flags`.

Current default modules: 
```cpp
#define HAL_MODULE_ENABLED
#define HAL_ADC_MODULE_ENABLED
#define HAL_CRC_MODULE_ENABLED //humm.. I think we don't use it and can be disabled 
#define HAL_DMA_MODULE_ENABLED
#define HAL_GPIO_MODULE_ENABLED
#define HAL_I2C_MODULE_ENABLED
#define HAL_PWR_MODULE_ENABLED
#define HAL_RCC_MODULE_ENABLED
#define HAL_SPI_MODULE_ENABLED
#define HAL_TIM_MODULE_ENABLED
#define HAL_USART_MODULE_ENABLED
#define HAL_CORTEX_MODULE_ENABLED
```

Optional "common" modules:
```cpp
#define HAL_PCD_MODULE_ENABLED   // for USB serial
#define HAL_HCD_MODULE_ENABLED  // for USB HOST (usb flash drive support)
#define HAL_SD_MODULE_ENABLED     // SDIO 
#define HAL_SRAM_MODULE_ENABLED  // to use FSMC
#define HAL_UART_MODULE_ENABLED   // for "old/normal" uart
```

### 3. `ldscript.ld`

We don't need it too. All boards with same mcu will have the same linker script. If we needed change the vector table offset, it could be done by the option `board_build.offset   = 0x8000`.

All other parameters comes from `board.json` and are equal for each mcu.

Parameters needed by the linker script:

- **LD_MAX_DATA_SIZE** - The size of the flash. It comes from `board.json`
- **LD_FLASH_OFFSET** - The offset from flash that the marlin firmware will be stored. It is the same as `VECT_TAB_OFFSET`. For example: if marlin should be stored in `0x08008000` and the `VECT_TAB_OFFSET` is `0x8000`, this value is `0x8000`. It comes from `board_build.offset ` in the board env.
- **LD_MAX_SIZE** - Ram size. It comes from the `board.json`

To use our generic variant, we need to add this script to the env: `buildroot/share/PlatformIO/scripts/stm32_bootloader.py`.

### 4. `PeripheralPins.c`

We just need edit it if and only if the board uses peripherals in "non common" pins. So, when we need a custom `PeripheralPins.c` we should know and write somewhere: "We needed this because our SPI FOO is linked to PBxx that is not configured in the generic variant".

### 5. `PinNamesVar.h`

This file is exactly the same for a mcu.

### 6. `variant.cpp`

The generic variant allow us to override about all parameters for this file - timers, oscillator value (hse value), SDA/SCL, etc...

If there's any needed for new overrides, we can just add it. About some weeks ago I added the option to override SDA/SCL, for example.

### 7. `variant.h`

Same for `variant.cpp`.


### 8. Example

```ini
#
# Board FOO BAR
#
[env:my_foo_bar_board]
platform             = ${common_stm32.platform}
extends              = common_stm32
build_flags          = ${common_stm32.build_flags} -DHAL_PCD_MODULE_ENABLED -DUSBCON -DUSBD_USE_CDC
board                  = genericSTM32F407VGT6
board_build.core     = stm32
board_build.variant  = MARLIN_F4x7Vx
board_build.ldscript = ldscript.ld
board_build.firmware = firmware.bin
board_build.offset   = 0x8000
board_upload.offset_address = 0x08008000
extra_scripts        = ${common.extra_scripts}
  pre:buildroot/share/PlatformIO/scripts/generic_create_variant.py
  buildroot/share/PlatformIO/scripts/stm32_bootloader.py
```

The main issue is: the only person capable of doing this job is the one that has the hardware on hands....

**Comment by thinkyhead** [comment]
I assume that between `THE_BOARD.json` and everything inside the `variants/THE_VARIANT` folder we have all the information we need. It should have come from the board developer or from someone else who knew the right values to apply.

**Comment by thisiskeithb** [comment]
> We should stop copying and pasting a bunch of 99% equal files for each added board... this is non sense and will lead to state we can call: variant hell...

100% agree, but as you’ve found out though trying to get the SKR 2 to use the generic environment, it’s not always that straight forward.

I’m sure we can get there, but I’d recruit testers of the generic variant before calling it good to go when new boards are added.

**Comment by rhapsodyv** [comment]
> > We should stop copying and pasting a bunch of 99% equal files for each added board... this is non sense and will lead to state we can call: variant hell...
> 
> 100% agree, but as you’ve found out though trying to get the SKR 2 to use the generic environment, it’s not always that straight forward.
> 
> I’m sure we can get there, but I’d recruit testers of the generic variant before calling it good to go when new boards are added.

Sure. But I think @bigtreetech can help us :-)

**Comment by thisiskeithb** [comment]
> Sure. But I think @bigtreetech can help us :-)

True! There's a regular group of beta board testers that I have access to as well.

**Comment by thinkyhead** [comment]
I have a few donated BTT and MKS boards on hand, and I can facilitate remote connections to these boards with a pretty fast VNC connection, if anyone wants to mess with one.

**Comment by trengtor** [comment]
Unification is undoubtedly good for the cause. But will it be convenient for modifying pins, etc. Plus bug fixing, which also takes time. In addition, the last release was six months old ...

**Comment by printercnc** [comment]
are topics like this still active? i just downloaded the latest firmware, tried to compile on BOARD_BLACK_STM32F407VE, reading the sd card on the board doesn't work. the sd card is still connected to the computer to save the gcode files, but the onboard port doesn't read the card, on the reprap lcd it says ATTACH MEDIA.

**Comment by thisiskeithb** [comment]
> are topics like this still active? i just downloaded the latest firmware, tried to compile on BOARD_BLACK_STM32F407VE, reading the sd card on the board doesn't work. the sd card is still connected to the computer to save the gcode files, but the onboard port doesn't read the card, on the reprap lcd it says ATTACH MEDIA.

You would need to compile with `BOARD_BTT_E3_RRF`, not `BOARD_BLACK_STM32F407VE`.

---

This Issue Queue is for Marlin bug reports and development-related issues, and we prefer not to handle user-support questions here. (As noted on [this page](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/.github/contributing.md#i-dont-want-to-read-this-whole-thing-i-just-have-a-question).) For best results getting help with configuration and troubleshooting, please use the following resources:

- [Marlin Documentation](https://marlinfw.org) hosted on MarlinFW.org
- [Marlin Forum](https://forums.reprap.org/list.php?415) hosted on RepRap.org
- [Marlin Firmware](https://www.facebook.com/groups/1049718498464482/) Facebook Group 
- [Marlin Firmware for 3D Printers](https://www.facebook.com/groups/3Dtechtalk/) Facebook Group
- [Marlin Configuration](https://www.youtube.com/results?search_query=marlin+configuration) on YouTube
- [Marlin Firmware](https://discord.com/servers/marlin-firmware-461605380783472640) Discord server

**Pull request merged**